### PR TITLE
Rebased check-in of InteractionMapping<T> Update

### PR DESCRIPTION
**Pull request by SimonDarksideJ**
Overview
---
Rebased check-in of InteractionMapping<T> Update and added Base DeviceManager/Controller implementation from @StephenHodgson 

Changes
---
* Renamed and updated InteractionMapping to MixedRealityInteractionMapping<T>
* New IMixedRealityInteractionmapping Interface for collections of MixedRealityInteractionMapping<T>
* Updated and enhanced tests for MixedRealityInteractionMapping<T>
* Took PR from @stephenhodgson to create base classes for DeviceManagers and Controllers (All scaffolds are now empty)
* Implemented the SixDof class (also in another PR)
* Tested with visualizer - working

Known Issues
---
Input System breaks again on launching with a controller, @StephenHodgson to investigate and fix after merge

Comments
---
I've left in tests for the old Dictionary method for InteractionMappings to show the speed comparison between the Dictionary Method and the new InteractionMapping array (I think Stephen may have finally beaten me down), as well as adding comparative speed comparisons between dictionary and array.  There's not much of a difference, but enough to warrant the change.

These will be removed once approved prior to merge.

### CODE REVIEW

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityInteractionMapping.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Internal.Definitions.Devices
+{
+    /// <summary>
+    /// Maps the capabilities of controllers, linking the Physical inputs of a controller to a Logical construct in a runtime project<para/>
+    /// <remarks>One definition should exist for each physical device input, such as buttons, triggers, joysticks, dpads, and more.</remarks>
+    /// </summary>
+    [Serializable]
+    public class MixedRealityInteractionMapping<TReadingType> : IMixedRealityInteractionMapping
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityInteractionMapping.cs]: I was going to comment on changes to this file but the meta must have gotten destroyed/corrupted.  Is there a way to fix so the diff is correct?

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityInteractionMapping.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        /// <returns>The current reading for the mapping</returns>
+        public TReadingType GetValue() => currentReading;
+
+        /// <summary>
+        /// Get the value of the Interaction Mapping data
+        /// </summary>
+        /// <param name="reading">Output parameter for the reading value, returns the current reading for the mapping</param>
+        public void GetValue(out TReadingType reading)
+        {
+            reading = currentReading;
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityInteractionMapping.cs]: I'd like to keep the direct calls to do a speed test.

File: Assets/MixedRealityToolkit/_Core/Devices/MixedReality/HoloLensClickerController.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+using Microsoft.MixedReality.Toolkit.Internal.Definitions.Devices;
+using Microsoft.MixedReality.Toolkit.Internal.Definitions.Utilities;
+using Microsoft.MixedReality.Toolkit.Internal.Interfaces;
+using Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem;
+using System.Collections.Generic;
+
+namespace Microsoft.MixedReality.Toolkit.Internal.Devices.WindowsMixedReality
+{
+    public class HoloLensClickerController : BaseController
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Devices/MixedReality/HoloLensClickerController.cs]: We can probably remove this. All of the interaction source controllers are virtually the same to the interaction manager

File: Assets/MixedRealityToolkit/_Core/Devices/WindowsGaming/WindowsGamingDevice.cs [hunk 4]
```diff
@@ -1,412 +0,0 @@
...
-using UnityEngine.XR.WSA.Input;
-
-#if WINDOWS_UWP
-using Windows.Gaming.Input;
-#endif // WINDOWS_UWP
-
-namespace Microsoft.MixedReality.Toolkit.Internal.Devices.WindowsGaming
-{
-    // TODO - Implement
-    public class WindowsGamingDevice : IMixedRealityDevice
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Devices/WindowsGaming/WindowsGamingDevice.cs]: Did we want to completely remove?

Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Devices/WindowsGaming/WindowsGamingDevice.cs]: RENAMED

File: Assets/MixedRealityToolkit/_Core/Devices/MixedReality/HoloLensClickerController.cs (hunk 3 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Devices/MixedReality/HoloLensClickerController.cs]: Except the controller mapping ISN'T

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityInteractionMapping.cs (hunk 2 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityInteractionMapping.cs]: Then following this PR, create some unit tests to compare.  I compared before and it was comparable but easier to manage

File: Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityController.cs [hunk 5]
```diff
@@ -15,270 +14,453 @@
...
         {
+            var interactions = new List<IMixedRealityInteractionMapping>();
+
             InputAction[] inputActions = Managers.MixedRealityManager.Instance.ActiveProfile.InputActionsProfile.InputActions;
             if (inputActions == null)
             {
                 return;
             }
+            //Add the Controller Pointer
+            interactions.Add(new MixedRealityInteractionMapping<SixDof>(1, AxisType.SixDof, DeviceInputType.SpatialPointer, new InputAction(1, "Select"))); // Note will convert these lookups to indexes
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityController.cs]: We should probably be getting this data from Managers.MixedRealityManager.Instance.ActiveProfile.ControllersProfile.MixedRealityControllerMappingProfiles

Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityController.cs]: it does, that is what the "SetupFromMapping" function is for.  The defaults are only for when there are NO mappings at all configured

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityController.cs]: We wouldn't have any mappings configured if we're not using the input system correct?

File: Assets/MixedRealityToolkit/_Core/Devices/BaseController.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        public ControllerState ControllerState { get; protected set; }
+
+        /// <inheritdoc />
+        public Handedness ControllerHandedness { get; }
+
+        /// <inheritdoc />
+        public IMixedRealityInputSource InputSource { get; }
+
+        /// <inheritdoc />
+        public IMixedRealityInteractionMapping[] Interactions { get; internal set; }
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Devices/BaseController.cs]: should be protected set.

Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Devices/BaseController.cs]: Grr, sure.

File: Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityController.cs (hunk 5 above)
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Devices/MixedReality/WindowsMixedRealityController.cs]: Err, no. The input system is meant to be "optional" :P 
#FoolofaTook :P

File: Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityInteractionMapping.cs [hunk 7]
```diff
@@ -223,14 +257,23 @@ public void SetValue(SixDof newValue)
                 Debug.LogError("SetValue(SixDof) is only valid for AxisType.SixDoF InteractionMappings");
             }
 
-            Changed = transformData.Position != newValue.Position ||
-                      transformData.Rotation != newValue.Rotation;
+            Changed = (sixdofData != null) ? !sixdofData.Equals(newValue) : (newValue != null);
```
Comment (StephenHodgson) [inline Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityInteractionMapping.cs]: SixDof is a struct and can never be null.

Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityInteractionMapping.cs]: So, what do you want the changed statement to be?

Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Definitions/Devices/MixedRealityInteractionMapping.cs]: See the PR I just opened. It addresses that and some other things.

**Comment by SimonDarksideJ** [comment]
@StephenHodgson latest testing proved this path is not optimal, however there were other changes that were viable.

Closing this PR and creating others for those changes.

**Pull request closed without merging**